### PR TITLE
VGG16 as Functional Subclass Model and Closures Removed

### LINKED ISSUE #1429: Make VGG a Functional subclass model & remove closures
This should follow the example of #1412 and #1401, including

- Docstring updates
- Colab demonstrating that weight loading works (if pre-trained weights exist) and numerics are identical

### PR DESCRIPTION
**Pull request by srikesh-07**
# What does this PR do?

Fixes #1429 

This [Colab](https://colab.research.google.com/drive/16fSGV4cubwpAnde-cZJw_2K5MVn5yzB7?usp=sharing) helps to reproduce the same results from existing and new VGG16.

## Who can review?
@ianstenbit  @LukeWood @tanzhenyu

### CODE REVIEW

File: keras_cv/metrics/coco/numerical_tests/GenerateSamples.ipynb [hunk 1]
```diff
@@ -56,7 +56,7 @@
       "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib>=2.1.0->pycocotools==2.0) (1.3.2)\n",
       "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.1->matplotlib>=2.1.0->pycocotools==2.0) (1.15.0)\n",
       "Building wheels for collected packages: pycocotools\n",
-      "  Building wheel for pycocotools (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
+      "  Building wheel for pycocotools (setup.py) ... \u001B[?25l\u001B[?25hdone\n",
```
Comment (ianstenbit) [inline keras_cv/metrics/coco/numerical_tests/GenerateSamples.ipynb:59]: Please remove the changes to this notebook from this PR

File: keras_cv/models/vgg16.py [hunk 2]
```diff
@@ -25,35 +25,63 @@
...
-    include_top,
-    classes=None,
-    weights=None,
-    input_shape=(224, 224, 3),
-    input_tensor=None,
-    pooling=None,
-    classifier_activation="softmax",
-    name="VGG16",
-    **kwargs,
+def build_vgg_block(
```
Comment (ianstenbit) [inline keras_cv/models/vgg16.py:28]: nit: let's call this `apply_vgg_block` and call the first input `x`

File: keras_cv/models/vgg16.py [hunk 3]
```diff
@@ -73,93 +101,135 @@ def VGG16(
...
+        include_top=True,
+        classes=None,
+        weights=None,
+        input_shape=(224, 224, 3),
+        pooling=None,
+        classifier_activation="softmax",
+        name="VGG16",
+        **kwargs,
+    ):
+        self.include_rescaling = include_rescaling
```
Comment (ianstenbit) [inline keras_cv/models/vgg16.py:118]: References to `self` should come after the call to super().__init__ below

File: keras_cv/models/vgg16.py [hunk 4]
```diff
@@ -73,93 +101,135 @@ def VGG16(
...
+        return {
+            "include_rescaling": self.include_rescaling,
+            "include_top": self.include_top,
+            "name": self.name,
+            "input_shape": self.input_shape[1:],
+            "input_tensor": self.input_tensor,
+            "pooling": self.pooling,
+            "classes": self.classes,
+            "classifier_activation": self.classifier_activation,
+        }
```
Comment (ianstenbit) [inline keras_cv/models/vgg16.py:231]: Please include:
`"trainable": self.trainable` 
as well

**Review by ianstenbit — CHANGES_REQUESTED** [review]
Thanks for the PR!
The colab and the PR look good -- just a few small comments for you.

File: keras_cv/models/vgg16.py (hunk 4 above)
Comment (srikesh-07) [reply keras_cv/models/vgg16.py:231]: I have done all the changes. Kindly please review and let me know if any changes to be made.

File: keras_cv/models/vgg16.py [hunk 5]
```diff
@@ -25,35 +25,62 @@
...
+    num_layers,
+    kernel_size,
+    stride,
+    activation,
+    padding,
+    max_pool,
+    name,
 ):
+    """
+    Builds VGG block
```
Comment (ianstenbit) [inline keras_cv/models/vgg16.py:39]: nit: applies a VGG block

File: keras_cv/models/vgg16.py [hunk 6]
```diff
@@ -25,35 +25,62 @@
...
+    Args:
+        x: None (or) Tensor, input tensor to pass through network
```
Comment (ianstenbit) [inline keras_cv/models/vgg16.py:41]: nit: I don't think x can be `None` here -- only a Tensor

**Review by ianstenbit — APPROVED** [review]
Thank you -- looks good! I left 2 more minor comments and then we can merge.

**Review by srikesh-07 — COMMENTED** [review]
I have done the final changes as mentioned, Thank you.

File: keras_cv/models/vgg16.py (hunk 5 above)
Comment (srikesh-07) [reply keras_cv/models/vgg16.py]: Done.

File: keras_cv/models/vgg16.py (hunk 6 above)
Comment (srikesh-07) [reply keras_cv/models/vgg16.py]: Done.

**Comment by srikesh-07** [comment]
Recently found there is a mismatch in the variable name. So I have fixed it, Kindly please check and approve d7ac677

**Review by srikesh-07 — COMMENTED** [review]
Recently found there is a mismatch in the variable name. So I have fixed it, Kindly please check and approve https://github.com/keras-team/keras-cv/commit/d7ac677c8c6853763fd8df42f3c87dfd9b9be7fa @ianstenbit

**Comment by srikesh-07** [comment]
I have also completed subclasssing for VGG-19. Can I have the same PR else Can I create a separate PR ? @ianstenbit

**Comment by srikesh-07** [comment]
I have completed the sub classing for the VGG-19 also. Can I have VGG-19 in this same PR or Could I create a separate PR ?

**Comment by ianstenbit** [comment]
> I have completed the sub classing for the VGG-19 also. Can I have VGG-19 in this same PR or Could I create a separate PR ? @ianstenbit

Let's do a separate PR for VGG 19

**Review by ianstenbit — APPROVED** [review]
Thanks!

**Pull request closed without merging**